mixnet/node/src/lib.rs

```diff
@@ -77,205 +76,279 @@ impl MixnetNode {
-        body: Body,
-        to: NymNodeRoutingAddress,
-    ) -> Result<()> {
+    async fn forward(&self, msg: Body, to: NymNodeRoutingAddress) -> Result<()> {
         let addr = SocketAddr::from(to);
-        let arc_socket = pool.get_or_init(&addr).await?;
-
-        if let Err(e) = {
-            let mut socket = arc_socket.lock().await;
-            body.write(&mut *socket).await
-        } {
-            tracing::error!("Failed to forward packet to {addr} with error: {e}. Retrying...");
-            return mixnet_protocol::retry_backoff(
-                addr,
-                max_retries,
-                retry_delay,
-                body,
-                arc_socket,
-            )
-            .await
-            .map_err(Into::into);
-        }
+
+        self.message_tx.send(TargetedMessage::new(addr, msg))?;
         Ok(())
     }
 }
+
+struct MessageHandler {
+    // TODO: remove this allow when we implement the retry logic
+    #[allow(dead_code)]
+    config: MixnetNodeConfig,
+    message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+    message_tx: mpsc::UnboundedSender<TargetedMessage>,
+    connections: HashMap<SocketAddr, TcpStream>,
+}
+
+impl MessageHandler {
+    pub fn new(
+        message_tx: mpsc::UnboundedSender<TargetedMessage>,
+        message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+        config: MixnetNodeConfig,
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                msg = self.message_rx.recv() => {
+                    if let Some(msg) = msg {
+                        match self.send(msg).await {
+                            Ok(Some(msg)) => {
+                                let _ = self.message_tx.send(msg);
+                            }
+                            Ok(None) => {},
+                            Err(e) => {
+                                tracing::error!("failed to send msg: {e}");
+                            },
+                        }
+                    } else {
+                        // Channel closed, we should shutdown the message handler thread
+                        return;
+                    }
+                },
+                _ = tokio::signal::ctrl_c() => {
+                    tracing::info!("Shutting down message handler thread...");
+                    return;
+                }
+            }
+        }
+    }
+
+    /// Send a message to the remote node,
+    /// return Ok(Some((Duration, Message))) if the message is not sent and the error is retryable
+    /// return Ok(None) if the message is sent successfully
+    /// return Err(e) if the message is not sent and the error is not retryable
+    async fn send(&mut self, mut msg: TargetedMessage) -> Result<Option<TargetedMessage>> {
```

Comment:
While reading this function, several thoughts came to my mind. 

1. Don't we also need to retry the case when `TcpStream::connect(..)` is failed?

2. Shouldn’t we sleep before trying to re-establish the connection (, instead of sleeping after that)?

3. Could we make a sleep in a separated task, as Giacomo implemented in the libp2p network backend? For example, you may be able to spawn an one-time task who takes a sleep and re-pushes the message into the queue. Then, other messages in the queue could be processed faster.


What do you think?

Reply:
I guess, some of above is not easy to be implemented simply with only one single task without any lock.

Reply:
Yeah, I am also aware of this problem. `sleep` will block the thread, in [this branch](https://github.com/logos-co/nomos-node/tree/mixnet-fan-in-retry), I introduced a retry pool, which means we can implement batching retry logic and not block the current thread. But unsure if we want this implementation at the moment.

Reply:
Yeah, I am also aware of this problem. sleep will block the thread. For this problem, I thought about use a single task thread, but it seems that we have to bring `Arc<Mutex<TcpStream>>` back, then contention on the same TcpStream will show up again. In [this branch](https://github.com/logos-co/nomos-node/tree/mixnet-fan-in-retry), I introduced a retry pool, which means we can implement batching retry logic and not block the current thread. But unsure if we want this implementation at the moment.

Reply:
> sleep will block the thread

This sleep will block the *task* not the thread. It's an important difference.
However, it's not acceptable to have the sleep here as it means all other messages can't be forwarded.
Agree with what @youngjoon-lee said, you can pretty much copy what we did in the libp2p network backend